### ISSUE TITLE
s390x Debian packages have checksum mismatches with the actual files

### ISSUE THREAD

**Issue by andyedwardsibm** [docker/docker-ce-packaging#937]
When running `curl -fsSL get.docker.com | bash` on a Debian Linux s390x machine, the following errors are reported

```
# Executing docker install script, commit: c2de0811708b6d9015ed1a2c80f02c9b70c8ce7b
+ sh -c 'apt-get update -qq >/dev/null'
+ sh -c 'DEBIAN_FRONTEND=noninteractive apt-get install -y -qq apt-transport-https ca-certificates curl >/dev/null'
+ sh -c 'install -m 0755 -d /etc/apt/keyrings'
+ sh -c 'curl -fsSL "https://download.docker.com/linux/ubuntu/gpg" | gpg --dearmor --yes -o /etc/apt/keyrings/docker.gpg'
gpg: WARNING: unsafe permissions on homedir '/root/.gnupg'
+ sh -c 'chmod a+r /etc/apt/keyrings/docker.gpg'
+ sh -c 'echo "deb [arch=s390x signed-by=/etc/apt/keyrings/docker.gpg] https://download.docker.com/linux/ubuntu focal stable" > /etc/apt/sources.list.d/docker.list'
+ sh -c 'apt-get update -qq >/dev/null'
+ sh -c 'DEBIAN_FRONTEND=noninteractive apt-get install -y -qq docker-ce docker-ce-cli containerd.io docker-compose-plugin docker-ce-rootless-extras docker-buildx-plugin >/dev/null'
E: Failed to fetch https://download.docker.com/linux/ubuntu/dists/focal/pool/stable/s390x/docker-buildx-plugin_0.11.1-1~ubuntu.20.04~focal_s390x.deb  Hash Sum mismatch
   Hashes of expected file:
    - SHA512:a23a1f56d0ddfbe488579abce934f62125b13835dbe89e85305e3876be165508ae1a59e46b6dc0c1345d4eea3850cafbe596801fa073e89cb8a8aba055055238
    - SHA256:ea91370d109896350d9731343328f70ea07cfac15136d7c7762864cc8418219a
    - SHA1:056cedad134033e5e7608779b87fdabee4011e92 [weak]
    - MD5Sum:a7e696fdb9a8d27a6747ed260038358b [weak]
    - Filesize:27160618 [weak]
   Hashes of received file:
    - SHA512:97a00b2e99b5388618c96b829410ac6ddf35a2a788b838bb80cad3e3bdd798c7b3645f2bceda9bc305227136b2f3d013e281c4f3d0132dc8f75dc85725b5e31a
    - SHA256:c466da225a26b3ed02a3876e85ed28c48b9d8c48bb32d2e947156375a6ccfd7f
    - SHA1:0a70e1ae882eafce3b62e8228411f308387bbc21 [weak]
    - MD5Sum:f6bd0c1c30201ca753843bf3c661d79a [weak]
    - Filesize:27160618 [weak]
   Last modification reported: Fri, 25 Aug 2023 16:05:46 +0000
E: Failed to fetch https://download.docker.com/linux/ubuntu/dists/focal/pool/stable/s390x/docker-ce-cli_24.0.3-1~ubuntu.20.04~focal_s390x.deb  Hash Sum mismatch
   Hashes of expected file:
    - SHA512:441606cc8fe003db41ac4847ad754a760ddf511dde9e96bf4e5f3a48391925abfbb8d2da2a4a5f79c0435436f09a803e1e8d4e92a82dac67a97ec54712d22307
    - SHA256:c14cdf2f87658a2c7a237c569d143c98c6e77fffe6706ff6be2c7dd080ec3b82
    - SHA1:94843c1f452e57e4c65876f1ca738dada2574682 [weak]
    - MD5Sum:9779e3ea2a1eea2a7f1e9846538c7611 [weak]
    - Filesize:12625210 [weak]
   Hashes of received file:
    - SHA512:b7ef1c9408f226279654d7c5ea2f47825a771f5623c433acb9c4ba14b2f8071be0416b101e5bb4fd8dc76b19958fc16e1a098c490bf4f663a345636de696feca
    - SHA256:be48f6256c4a8e18bc342fde6b99bc996abb85354bfe1bcf9bd0805e70583ea5
    - SHA1:35354bb60172a26bb75cb20f1db18a1464ef7b48 [weak]
    - MD5Sum:b68479806729e67281ca4d103a2db586 [weak]
    - Filesize:12625210 [weak]
   Last modification reported: Fri, 25 Aug 2023 16:05:54 +0000
E: Failed to fetch https://download.docker.com/linux/ubuntu/dists/focal/pool/stable/s390x/docker-ce_24.0.3-1~ubuntu.20.04~focal_s390x.deb  Hash Sum mismatch
   Hashes of expected file:
    - SHA512:bf96935b6db864510a95e1842cc54b956c3ed6175a56fa9244dccf00b323b9397f0903216d2497ca2982b7c93f7e46d124801fc190d9640476259a9178c1b403
    - SHA256:6d1c1656d82b621fcec013566ddb67537bf53cb73ca3ae839fbbd83804b93796
    - SHA1:fd5b0d5a95af5a6459fb120362a84edd3a7ffe38 [weak]
    - MD5Sum:df3e70ebf3512c73d2b553c145991652 [weak]
    - Filesize:14521578 [weak]
   Hashes of received file:
    - SHA512:d261443560924d398d311213c966b26f8361663c5dd15f133cb109b5c0d555ca17effd8d550d41d4cfb39315b96e416cbdc5da9f172f1447624f66aa0c30b744
    - SHA256:49af8713d2866fa219d639e615b2fa425aad55c682ef9e1aa084b908f4633bf7
    - SHA1:aa6b5d1aaf649d2eeefce87623fed8b687665f00 [weak]
    - MD5Sum:f0032cda2c98e313340e838a27b278e8 [weak]
    - Filesize:14521578 [weak]
   Last modification reported: Fri, 25 Aug 2023 16:06:01 +0000
E: Failed to fetch https://download.docker.com/linux/ubuntu/dists/focal/pool/stable/s390x/docker-ce-rootless-extras_24.0.3-1~ubuntu.20.04~focal_s390x.deb  Hash Sum mismatch
   Hashes of expected file:
    - SHA512:108ea31b51372fbb3dca683486c81149ea72aab0a1c5e86f06913e4cdfc61cd76a707501da1778f1ea4feb395aca8cf205ec0886b0d0450805c7d381413f7f1d
    - SHA256:752eacbeb382f7fe77f28a43bc61af47126e8cf4a125b98f6064f6e962ac244c
    - SHA1:0c48e0931a0bef3dab5f8ece4fcf863b9578233c [weak]
    - MD5Sum:852a98266cf76e77d164fd241a50d345 [weak]
    - Filesize:8585952 [weak]
   Hashes of received file:
    - SHA512:3acb8614dfae3c739f38c312969a7e479cd6da869e1642d065fe07dd5ec9cdbeda86e959d602c790455679982b53c4ef7e797e8ebfef0ecf343118a0a5686924
    - SHA256:d5ed2158915f6284429f44c61ac6e491ce07bffa1d532eb67548cc33c6db1d60
    - SHA1:1b9bb9163773d270745a246335ba003a172fe911 [weak]
    - MD5Sum:9e02e6164f5654a3862a18616da607f1 [weak]
    - Filesize:8585952 [weak]
   Last modification reported: Fri, 25 Aug 2023 16:05:57 +0000
E: Failed to fetch https://download.docker.com/linux/ubuntu/dists/focal/pool/stable/s390x/docker-compose-plugin_2.19.1-1~ubuntu.20.04~focal_s390x.deb  Hash Sum mismatch
   Hashes of expected file:
    - SHA512:ada47efffbda879c50a15c7a64c8b382fe5f4afd985e86e4ee529694458abc0b68b4ab679271606bd69720bb7a845b6230d5fbc42a10acec1fd2f6607bf8a443
    - SHA256:5baf45c090caaca6e6b43b1258d654e397ebfd454e0228ee72cb485e60264065
    - SHA1:0d2e1a48c6b3b7b2d6a32223aa76725067f5d349 [weak]
    - MD5Sum:b2737448590743d57c74d51204f9e54a [weak]
    - Filesize:10734326 [weak]
   Hashes of received file:
    - SHA512:46883156f2b0176e5cda69c11cdefda2dc9272e86c2ec11d5b8f42103605c87368cb0a8cfed3982c2f2ff80c466446e118e51c1dc908e4a9a3f7d878af049af2
    - SHA256:99205c1c2bb723d7a2666a01fcd8b3f383f81d9ea808e5223a24d74f1025212b
    - SHA1:f9d39bb74b918c1c2a16a80c33da8521fab5baa8 [weak]
    - MD5Sum:03a46ad6d138e787b6c842f1265b99f9 [weak]
    - Filesize:10734326 [weak]
   Last modification reported: Fri, 25 Aug 2023 16:06:03 +0000
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
```

Considering https://download.docker.com/linux/ubuntu/dists/focal/pool/stable/s390x/docker-ce_24.0.3-1~ubuntu.20.04~focal_s390x.deb for the core Docker package, I can see that https://download.docker.com/linux/ubuntu/dists/focal/stable/binary-s390x/Packages contains the following entry:

```
Package: docker-ce
Architecture: s390x
Version: 5:24.0.3-1~ubuntu.20.04~focal
Priority: optional
Section: admin
Maintainer: Docker <support@docker.com>
Installed-Size: 76693
Depends: containerd.io (>= 1.6.4), docker-ce-cli, iptables, libseccomp2 (>= 2.3.0), libc6 (>= 2.8), libdevmapper1.02.1 (>= 2:1.02.97), libsystemd0
Recommends: apparmor, ca-certificates, docker-ce-rootless-extras, git, libltdl7, pigz, procps, xz-utils
Suggests: cgroupfs-mount | cgroup-lite
Conflicts: docker (<< 1.5~), docker-engine, docker-engine-cs, docker.io, lxc-docker, lxc-docker-virtual-package
Replaces: docker-engine
Filename: dists/focal/pool/stable/s390x/docker-ce_24.0.3-1~ubuntu.20.04~focal_s390x.deb
Size: 14521578
MD5sum: df3e70ebf3512c73d2b553c145991652
SHA1: fd5b0d5a95af5a6459fb120362a84edd3a7ffe38
SHA256: 6d1c1656d82b621fcec013566ddb67537bf53cb73ca3ae839fbbd83804b93796
SHA512: bf96935b6db864510a95e1842cc54b956c3ed6175a56fa9244dccf00b323b9397f0903216d2497ca2982b7c93f7e46d124801fc190d9640476259a9178c1b403
Homepage: https://www.docker.com
Description: Docker: the open-source application container engine
 Docker is a product for you to build, ship and run any application as a
 lightweight container
 .
 Docker containers are both hardware-agnostic and platform-agnostic. This means
 they can run anywhere, from your laptop to the largest cloud compute instance and
 everything in between - and they don't require you to use a particular
 language, framework or packaging system. That makes them great building blocks
 for deploying and scaling web apps, databases, and backend services without
 depending on a particular stack or provider.
```

But the checksums for the file at https://download.docker.com/linux/ubuntu/dists/focal/pool/stable/s390x/docker-ce_24.0.3-1~ubuntu.20.04~focal_s390x.deb don't match those values - they are...

```
andy:foo$ md5sum docker-ce_24.0.3-1~ubuntu.20.04~focal_s390x.deb 
f0032cda2c98e313340e838a27b278e8  docker-ce_24.0.3-1~ubuntu.20.04~focal_s390x.deb
andy:foo$ sha1sum docker-ce_24.0.3-1~ubuntu.20.04~focal_s390x.deb 
aa6b5d1aaf649d2eeefce87623fed8b687665f00  docker-ce_24.0.3-1~ubuntu.20.04~focal_s390x.deb
andy:foo$ sha256sum docker-ce_24.0.3-1~ubuntu.20.04~focal_s390x.deb 
49af8713d2866fa219d639e615b2fa425aad55c682ef9e1aa084b908f4633bf7  docker-ce_24.0.3-1~ubuntu.20.04~focal_s390x.deb
andy:foo$ sha512sum docker-ce_24.0.3-1~ubuntu.20.04~focal_s390x.deb 
d261443560924d398d311213c966b26f8361663c5dd15f133cb109b5c0d555ca17effd8d550d41d4cfb39315b96e416cbdc5da9f172f1447624f66aa0c30b744  docker-ce_24.0.3-1~ubuntu.20.04~focal_s390x.deb
```

This is true for multiple s390x files and at least the following:
- https://download.docker.com/linux/ubuntu/dists/focal/pool/stable/s390x/docker-buildx-plugin_0.11.1-1~ubuntu.20.04~focal_s390x.deb
- https://download.docker.com/linux/ubuntu/dists/focal/pool/stable/s390x/docker-ce-cli_24.0.3-1~ubuntu.20.04~focal_s390x.deb
- https://download.docker.com/linux/ubuntu/dists/focal/pool/stable/s390x/docker-ce_24.0.3-1~ubuntu.20.04~focal_s390x.deb
- https://download.docker.com/linux/ubuntu/dists/focal/pool/stable/s390x/docker-ce-rootless-extras_24.0.3-1~ubuntu.20.04~focal_s390x.deb
- https://download.docker.com/linux/ubuntu/dists/focal/pool/stable/s390x/docker-compose-plugin_2.19.1-1~ubuntu.20.04~focal_s390x.deb

All of these seem to have been published as a set on 2023-08-31 at around 16:49:30.  I've not checked the files for any of the other Ubuntu versions, or any other Linux distro.  We've _not_ seen failures on any other architecture though, so it looks like it's just some set of files for s390x

**Comment by mknight-atl** [comment]
We also observed hash sum mismatches on the following packages that we use:
```
# ARM64
/dists/jammy/pool/stable/arm64/docker-ce-cli_24.0.6-1~ubuntu.22.04~jammy_arm64.deb
/dists/jammy/pool/stable/arm64/docker-ce_24.0.6-1~ubuntu.22.04~jammy_arm64.deb
/dists/jammy/pool/stable/arm64/docker-ce-rootless-extras_24.0.6-1~ubuntu.22.04~jammy_arm64.deb
/dists/jammy/pool/stable/arm64/docker-compose-plugin_2.21.0-1~ubuntu.22.04~jammy_arm64.deb
# AMD64
/dists/jammy/pool/stable/amd64/docker-ce-cli_24.0.6-1~ubuntu.22.04~jammy_amd64.deb
/dists/jammy/pool/stable/amd64/docker-ce_24.0.6-1~ubuntu.22.04~jammy_amd64.deb
/dists/jammy/pool/stable/amd64/docker-ce-rootless-extras_24.0.6-1~ubuntu.22.04~jammy_amd64.deb
/dists/jammy/pool/stable/amd64/docker-compose-plugin_2.21.0-1~ubuntu.22.04~jammy_amd64.deb
```
Diffing the files confirmed that the same versions had been rebuilt and pushed. For example looking, at part of a binary diff of `docker-compose-plugin_2.21.0-1~ubuntu.22.04~jammy_arm64.deb`:
```
640736,640737c640736,640737
< 009c6df0: 4461 7465 3a20 5475 6520 5365 7020 2035  Date: Tue Sep  5
< 009c6e00: 2031 363a 3532 3a33 3220 3230 3233 0a52   16:52:32 2023.R
---
> 009c6df0: 4461 7465 3a20 4672 6920 5365 7020 2038  Date: Fri Sep  8
> 009c6e00: 2031 373a 3031 3a32 3120 3230 3233 0a52   17:01:21 2023.R
```
There were other minor differences in what looked to be a PGP cert inside the file.

We had to ask our package mirror maintainers to manually delete these files from their cache to fix our builds.

My understanding is that pushing up a different build of the same version is considered bad practice. Is there any reason why these weren't released as new packages with their build number bumped (e.g. `24.0.6-2`)?

**Comment by andyedwardsibm** [comment]
Okay, so someone's done something because today...

```
andy:foo$ wget https://download.docker.com/linux/ubuntu/dists/focal/pool/stable/s390x/docker-ce_24.0.3-1~ubuntu.20.04~focal_s390x.deb

...

andy:foo$ md5sum docker-ce_24.0.3-1~ubuntu.20.04~focal_s390x.deb 
df3e70ebf3512c73d2b553c145991652  docker-ce_24.0.3-1~ubuntu.20.04~focal_s390x.deb
andy:foo$ sha1sum docker-ce_24.0.3-1~ubuntu.20.04~focal_s390x.deb 
fd5b0d5a95af5a6459fb120362a84edd3a7ffe38  docker-ce_24.0.3-1~ubuntu.20.04~focal_s390x.deb
andy:foo$ sha256sum docker-ce_24.0.3-1~ubuntu.20.04~focal_s390x.deb 
6d1c1656d82b621fcec013566ddb67537bf53cb73ca3ae839fbbd83804b93796  docker-ce_24.0.3-1~ubuntu.20.04~focal_s390x.deb
andy:foo$ sha512sum docker-ce_24.0.3-1~ubuntu.20.04~focal_s390x.deb 
bf96935b6db864510a95e1842cc54b956c3ed6175a56fa9244dccf00b323b9397f0903216d2497ca2982b7c93f7e46d124801fc190d9640476259a9178c1b403  docker-ce_24.0.3-1~ubuntu.20.04~focal_s390x.deb
```

So now all the checksums match up with what the `Packages` files claim

Looking at https://download.docker.com/linux/ubuntu/dists/focal/pool/stable/s390x/, the timestamps on the files are unchanged

**Comment by neersighted** [comment]
Ah, there are two issues here, both of which are related to the same event. I wrote a bit about it over here: https://github.com/moby/moby/issues/46470#issuecomment-1718230952

In short, the metadata for IBM packages was out of sync with the package files, and there were limited options to recover; the "least evil" was to re-sign the 24.0.6 binaries for non-IBM platforms (currently, signing and publishing are the same step in the release pipeline).

This situation should happen rarely or not at all; this infrastructure is being actively worked on and I don't think we'll have to repeat this, but I'm also looking at e.g. separating signing and publishing to prevent this scenario more robustly in the future.

I wrote a couple more thoughts on mirrors over at https://github.com/docker/docker-install/issues/379#issuecomment-1727818067.

As a final take-away, it seems like we'll really need to figure out a canonical venue to discuss Docker CE/download.docker.com since issues end up opened in so many diverse places (and I also receive emails, Slack messages, etc.); I'll continue to mull over where that should be, as the canonical pipeline that defines this infrastructure is actually 4-5 internal repositories.
